Source/Scene/PointCloudPostProcessor.js

```diff
@@ -0,0 +1,1109 @@
+                pixelFormat : PixelFormat.RGBA,
+                pixelDatatype : PixelDatatype.UNSIGNED_BYTE,
+                sampler : createSampler()
+            });
+        }
+
+        // There used to be an explanation of how this worked here
+        // but it got too long.
+        // TODO: Find a better place to put an explanation of what all
+        // the framebuffers are meant for.
+        processor._framebuffers = {
+            prior : new Framebuffer({
+                context : context,
+                colorTextures : [
+                    colorTextures[0],
+                    ecTexture
+                ],
+                depthStencilTexture : dirty,
+                destroyAttachments : false
+            }),
+            screenSpacePass : new Framebuffer({
+                context : context,
+                colorTextures : [depthTextures[0], aoTextures[0]],
+                depthStencilTexture: dirty,
+                destroyAttachments : false
+            }),
+            aoBufferA : new Framebuffer({
+                context : context,
+                colorTextures : [aoTextures[1]],
+                destroyAttachments : false
+            }),
+            aoBufferB : new Framebuffer({
+                context : context,
+                colorTextures : [aoTextures[0]],
+                destroyAttachments : false
+            }),
+            stencilMask : new Framebuffer({
+                context : context,
+                depthStencilTexture: dirty,
+                destroyAttachments : false
+            }),
+            densityEstimationPass : new Framebuffer({
+                context : context,
+                colorTextures : [densityMap],
+                depthStencilTexture: dirty,
+                destroyAttachments : false
+            }),
+            regionGrowingPassA : new Framebuffer({
+                context : context,
+                colorTextures : [colorTextures[1],
+                                 depthTextures[1],
+                                 aoTextures[1]],
+                depthStencilTexture: dirty,
+                destroyAttachments : false
+            }),
+            regionGrowingPassB : new Framebuffer({
+                context: context,
+                colorTextures: [colorTextures[0],
+                                depthTextures[0],
+                                aoTextures[0]],
+                depthStencilTexture: dirty,
+                destroyAttachments: false
+            })
+        };
+        processor._depthTextures = depthTextures;
+        processor._densityTexture = densityMap;
+        processor._edgeCullingTexture = edgeCullingTexture;
+        processor._sectorLUTTexture = sectorLUTTexture;
+        processor._aoTextures = aoTextures;
+        processor._colorTextures = colorTextures;
+        processor._ecTexture = ecTexture;
+        processor._dirty = dirty;
+    }
+
+    function replaceConstants(sourceStr, constantName, replacement) {
```

Comment:
So, just to clarify, you're not suggesting using ShaderSource's `defines` array, and instead constructing the shader string on the fly?

Reply:
Normally I would say add them to the `defines` array, but since they are setting constants it's different than what `defines` is traditionally used for. So yeah, adding the defines to the shader string is the right approach for now.

Reply:
The problem with just appending the `#define`s to the shader string is that it's not clear where the new lines should be appended. They can't be appended at the start because [GLSL requires](https://www.khronos.org/opengl/wiki/Core_Language_(GLSL)#Extensions) `#extension`s to be listed right after the version declaration (i.e so they have to be appended after the extension declaration(s) but before any uniform declarations).

Ideally we'd solve this problem by working with `ShaderSource` objects rather than just shader strings. Unfortunately, we also want to use the `createQuadViewportFS` method from `Context`, which doesn't take a ShaderSource.

Reply:
The `#extensions` shouldn't be a problem as of the GLSL Modernizer PR:  https://github.com/AnalyticalGraphicsInc/cesium/pull/5623